force-app/main/default/classes/ContactAccountHandler.cls

```diff
@@ -0,0 +1,88 @@
+                        if (contact.Is_Primary__c) {
+                            primaryContactCount++;
+                        }
+                    }
+                    if (primaryContactCount > 1) {
+                        contactMap.get(account.Id).addError('Primary Contact Already Exists');
+                    }
+                }
+            } 
+        }
+    
+        public static void setParentAccountsEmail(List<Contact> contactList){
+            Set<Id> accountIdSet = new Set<Id>();
+            for (Contact contact : contactList) {
+                if (contact.Is_Primary__c) {
+                    //add contact account id if contact is primary contact.
+                    accountIdSet.add(contact.AccountId);
+                }
+            }
+            List<Account> accountList = [Select Id, Name, ParentId, (Select Id, Name, Email, Is_Primary__c from Contacts) from Account where Id in :accountIdSet];
+            //get all accounts
+            Map<Id, Account> accountMap = new Map<Id, Account>();
+            for (Account account : accountList) {
+                accountMap.put(account.Id, account);
+            }
+            //get parent account
+            Set<Id> parentAccountIdSet = new Set<Id>();
+            for (Account account : accountList) {
+                parentAccountIdSet.add(account.ParentId);
+            }
+    
+            List<Contact> parentAccountContactList = [Select Id, AccountId, Email, Name, Is_Primary__c from Contact where AccountId in :parentAccountIdSet and Is_Primary__c = true];
+            //get all contacts of parent accounts
+            Map<Id, Contact> parentContactMap = new Map<Id, Contact>();
+            for (Contact contact : parentAccountContactList) {
+                parentContactMap.put(contact.AccountId, contact);
+            }
+    
+            List<Contact> updatedContactList = new List<Contact>();
+            //setting the email of parents accounts primary contacts email
+            for (Contact contact : contactList) {
+                if (accountMap.containsKey(contact.AccountId) && parentContactMap.containsKey(accountMap.get(contact.AccountId).ParentId)) {
+                    Contact parentContact = parentContactMap.get(accountMap.get(contact.AccountId).ParentId);
+                    parentContact.Email = contact.Email;
+    
+                    updatedContactList.add(parentContact);
+                }
+            }
+    
+            if (updatedContactList.size() > 0) {
+                update updatedContactList;
+            }
+        }
+    
+        public static void getContactsWithAccount(List<Contact> contactList){
+            Set<Id> accountIdSet = new Set<Id>();
+            for (Contact contact : contactList) {
+                if (contact.AccountId != null) {
+                    accountIdSet.add(contact.AccountId);
+                }
+            }
+            List<Account> accountList = [Select Id, Name From Account where Id in :accountIdSet];
```

Comment:
we can also perform an inner query for this requirement, also let's discuss a way to archive this requirement by admin configuration.